main/lsp/requests/workspace_symbols.cc

```diff
@@ -21,12 +44,134 @@ unique_ptr<SymbolInformation> symbolRef2SymbolInformation(const LSPConfiguration
     if (location == nullptr) {
         return nullptr;
     }
-    auto result =
-        make_unique<SymbolInformation>(sym->name.show(gs), symbolRef2SymbolKind(gs, symRef), std::move(location));
+    auto result = new SymbolInformation(sym->name.show(gs), symbolRef2SymbolKind(gs, symRef), std::move(location));
     result->containerName = sym->owner.data(gs)->showFullName(gs);
     return result;
 }
 
+namespace {
+
+/** Returns a pair of {score, query_length_matched} for the given symbol/query. */
+pair<int, int> scoreSymbol(const string_view symbol, const string_view query) {
+    auto symbolIter = symbol.begin();
+    auto symbolEnd = symbol.end();
+    bool atBoundary = true;
+    int score = symbol.length();
+    int queryMatchLength = 0;
+    for (auto queryCh : query) {
+        int gap = 0;
+        int symbolCharsConsumed = 0;
+        bool queryCharIsLower = islower(queryCh);
+        while (symbolIter != symbolEnd) {
+            char symbolCh = *symbolIter++;
+            symbolCharsConsumed++;
+            // Check if the current character is boundary-starting
+            atBoundary = atBoundary || isupper(symbolCh);
+            if (queryCh == symbolCh || (queryCharIsLower && tolower(queryCh) == tolower(symbolCh))) {
+                queryMatchLength++;
+                if (gap == 0) {
+                    // no penalty for adjacent matching characters
+                } else if (atBoundary) {
+                    score += 100 + gap; // being *on* a boundary is good
+                } else {
+                    score += 200 + 5 * gap; // being *off* a boundary is bad
+                }
+                break;
+            }
+            atBoundary = !isalnum(symbolCh);
+            gap++;
+        }
+        if (queryMatchLength > 1) {
+            score += symbolCharsConsumed;
+        }
+        if (symbolIter == symbolEnd) {
+            break;
+        }
+    }
+    return make_pair(score, queryMatchLength);
+}
+} // namespace
+
+vector<unique_ptr<SymbolInformation>> SymbolMatcher::doQuery(string_view query_view, int limit) {
+    vector<unique_ptr<SymbolInformation>> results;
+    string query(query_view);
+    if (query.empty()) {
```

Comment:
(minor nit) you could probably make the same check on `query_view` before making the string.

Reply:
Excellent point.  In light of this and other comments about how I'm handling strings, am refactoring to avoid string construction, and this `empty` call should get refactored away.